### ISSUE TITLE
Broken on Ubuntu/Linux

### ISSUE THREAD

**Issue by michael-nischt** [atom/ide-csharp#7]
I tested ide-csharp on Windows 10 where it works just fine but on Ubuntu/Linux it doesn't work. In both cases I used Atom 1.21.0 x64

Maybe the error log helps:

```
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) Server starting "/home/micha/Projects/PwnieYard/unity/"
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) starting child Node "/home/micha/.atom/packages/ide-csharp/node_modules/omnisharp-client/languageserver/server.js"
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 Go (go-langserver) Server starting "/home/micha/Projects/PwnieYard/go/"
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.sendRequest initialize sending Object
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 Go (go-langserver) rpc.sendRequest initialize sending Object
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 Go (go-langserver) rpc.sendRequest initialize received (165ms) Object
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 Go (go-langserver) rpc.sendNotification textDocument/didOpen Object
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 Go (go-langserver) Server started "/home/micha/Projects/PwnieYard/go/" (pid 301)
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]
```

**Comment by rsese** [comment]
> module.js:470 Uncaught (in promise) Error: Cannot find module '../node_modules/nsfw/node_modules/fs-extra/lib/index.js' at Module._resolveFilename 

Hmmm /cc https://github.com/atom/atom/issues/15642?

Can you confirm if you have the same error in `1.22.0-beta0` or if things work ok there?

**Comment by michael-nischt** [comment]
There errors seem to be gone in 1.22.0-beta0 (see log below). However, the ide-csharp package doesn't seem to have any effect at all. No linting / auto-completion /.., nothing really.

`<embedded>:118269 IntersectionObserver.observe(target): target element is not a descendant of root.
TreeView @ <embedded>:118269
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) Server starting "/home/micha/Projects/PY-unity/"
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) starting child Node "/home/micha/.atom/packages/ide-csharp/node_modules/omnisharp-client/languageserver/server.js"
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.sendRequest initialize sending Objectcapabilities: Object__proto__: ObjectprocessId: 5996rootPath: "/home/micha/Projects/PY-unity/"__proto__: Object__defineGetter__: __defineGetter__()arguments: nullcaller: nulllength: 2name: "__defineGetter__"__proto__: ()__defineSetter__: __defineSetter__()__lookupGetter__: __lookupGetter__()__lookupSetter__: __lookupSetter__()constructor: Object()hasOwnProperty: hasOwnProperty()isPrototypeOf: isPrototypeOf()propertyIsEnumerable: propertyIsEnumerable()toLocaleString: toLocaleString()toString: toString()arguments: nullcaller: nulllength: 0name: "toString"__proto__: ()valueOf: valueOf()get __proto__: __proto__()set __proto__: __proto__()
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "telemetry/event"1: Objectlength: 2__proto__: Array[0]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "telemetry/event"1: Objectlength: 2__proto__: Array[0]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "telemetry/event"1: Objectlength: 2__proto__: Array[0]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "telemetry/event"1: Objectlength: 2__proto__: Array[0]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "telemetry/event"1: Objectlength: 2__proto__: Array[0]
/home/micha/.atom/packages/ide-go/node_modules/atom-languageclient/build/lib/logger.js:27 C# (OmniSharp) rpc.onNotification Array[2]0: "window/logMessage"1: Objectlength: 2__proto__: Array[0]
`

**Comment by the-codelight** [comment]
Same over here, installation on linux mint 18.2.
atom version 1.21.0x64
php7.1

hope you can fix it guys :+1: 

module.js:470 Uncaught (in promise) Error: Cannot find module '../node_modules/nsfw/node_modules/fs-extra/lib/index.js'
    at Module._resolveFilename (module.js:470:15)
    at Module._resolveFilename (/usr/share/atom/resources/electron.asar/common/reset-search-paths.js:35:12)
    at Function.get_Module._resolveFilename (/usr/share/atom/resources/app/src/module-cache.js:354:58)
    at Module.require (file:///usr/share/atom/resources/app.asar/static/index.js:40:43)
    at require (internal/module.js:20:19)
    at customRequire (/usr/share/atom/resources/app/static/<embedded>:96:26)
    at get_fse (/usr/share/atom/resources/app/node_modules/nsfw/lib/src/index.js:17:80)
    at Object.buildNSFW (/usr/share/atom/resources/app/node_modules/nsfw/lib/src/index.js:88:16)
    at nsfw (/usr/share/atom/resources/app/node_modules/nsfw/lib/src/index.js:33:37)
    at NSFWBackend.<anonymous> (/usr/share/atom/resources/app/src/path-watcher.js:192:34)
    at Generator.next (<anonymous>)
    at step (/usr/share/atom/resources/app/src/path-watcher.js:5:279)
    at /usr/share/atom/resources/app/src/path-watcher.js:5:449
    at NSFWBackend.<anonymous> (/usr/share/atom/resources/app/src/path-watcher.js:5:105)
    at NativeWatcher.<anonymous> (/usr/share/atom/resources/app/src/path-watcher.js:326:32)
    at Generator.next (<anonymous>)
    at step (/usr/share/atom/resources/app/src/path-watcher.js:5:279)
    at /usr/share/atom/resources/app/src/path-watcher.js:5:449
    at NativeWatcher.<anonymous> (/usr/share/atom/resources/app/src/path-watcher.js:5:105)
    at NativeWatcher.onDidChange (/usr/share/atom/resources/app/src/path-watcher.js:358:18)
    at PathWatcher.onDidChange (/usr/share/atom/resources/app/src/path-watcher.js:566:37)
    at /usr/share/atom/resources/app/src/path-watcher.js:575:24
    at process._tickCallback (internal/process/next_tick.js:103:7)

**Comment by damieng** [comment]
The nsfw error is unrelated to atom-ide.

Can you set the debugging mode so we can figure out what's actually going on?

Open DevTools window and type into the console;

`atom.config.set('core.debugLSP', true)` 

Then restart atom and look again in devtools.

**Comment by michael-nischt** [comment]
@damieng I said above, errors seem to have disappeared for in 1.22.0-beta0 but it's not working at all. The log above has `atom.config.set('core.debugLSP', true)` set. None of you guys are testing ide-csharp on Linux? Btw. the ancient omnisharp is working fine except the long list of deprecation warnings..

**Comment by damieng** [comment]
Sorry, the logs looked like they were just for the nsfw errors which are unrelated. I've edited your messages to clean them up.

It looks like it's starting the Omnisharp server but is not getting a response to the initialize message.

I haven't tested ide-csharp on Linux - the cross-platform nature of node and limited resources means we often don't get to try every combo every release.

**Comment by michael-nischt** [comment]
Any update on this issue? I understand the resources are limited but it has been a while.. =)

**Comment by daviwil** [comment]
@michael-nischt @the-codelight can you let me know what version of Mono is installed on your system?  Try running `mono --version`:

```
Mono JIT compiler version 4.8.0 (Stable 4.8.0.520/8f6d0f6 Thu Sep 21 07:12:44 UTC 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       normal
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug 
        LLVM:          supported, not enabled.
        GC:            sgen

```

**Comment by michael-nischt** [comment]
`
Mono JIT compiler version 4.6.2 (Debian 4.6.2.7+dfsg-1ubuntu1)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          supported, not enabled.
	GC:            sgen
`

**Comment by daviwil** [comment]
Thanks @michael-nischt!  Looks like OmniSharp requires Mono 5.2.0, but there's a fix we can do on our side to make it so that you don't have to install it yourself.  Should have this fixed pretty soon.  Thanks for your patience!

**Comment by michael-nischt** [comment]
Thanks, that would be most awesome since the latest version with Ubuntu 17.10 is 4.6 and I'd rather not install the PPA to have the latest installed system-wide since I had bad experiences with it in the past.

**Comment by daviwil** [comment]
Yep, definitely understood.  I think this is one case where you'll be able to escape it :)

**Comment by daviwil** [comment]
Hit another issue with OmniSharp that delayed this fix, but we should be unblocked soon!

**Comment by daviwil** [comment]
Just shipped 0.6.1 which has the fix for this issue!  Give it a shot and let us know how it goes for you.

**Comment by michael-nischt** [comment]
@daviwil thanks a lot. it works now! 

(FYI: it took more than a minute, maybe 2, after the restart until error/auto completion worked. After successive restarts it only took a few seconds though. Probably omnisharp initializing something the first time.)

**Comment by HebaruSan** [comment]
Is this how it's supposed to work?

![image](https://user-images.githubusercontent.com/1559108/32678821-a0818836-c629-11e7-87f3-f9b7d33ded11.png)

The StatusSymbol function is in the list but way, way down (or it was, now that I try it again, the list stops at E); by default Atom is now trying to autocomplete every word to "accepted". When I open a block with {, it inserts "accepted" after it when I press enter. It does it again when I close the block with }.

![image](https://user-images.githubusercontent.com/1559108/32678898-fdc5b332-c629-11e7-873a-b57c015ef880.png)

So far this is more of a hindrance than a help.

**Comment by damieng** [comment]
@HebaruSan the spec isn't clear whether the filtering is supposed to be done in the client or by the server so there is some variance.  We're working on some better caching/filtering code for autocomplete that will allow us to filter these for ide-csharp.

**Comment by daviwil** [comment]
@michael-nischt Yep, OmniSharp gets downloaded the first time you load the ide-csharp package after an install/update.  I'm working on adding a status indicator which will make it more obvious when OmniSharp is fully started and ready to go.

Regarding completion filtering, VS Code definitely does its own filtering client-side.  Generally they seem to get the completions list for the first character you type and then continue filtering down the list based on the substring that you've typed.  As @damieng says, the improvements made to autocomplete will help address this.
